ColorPicker Update Roll-up

robloo opened the pull request:
## What does the pull request do?

 * Cherry-picks some commits from https://github.com/AvaloniaUI/Avalonia/pull/9074 which ran into some framework bugs that need fixes first
 * Fixes https://github.com/AvaloniaUI/Avalonia/issues/9098 ~by adding "a new property for `IsHostedInColorPicker`... [to] switch the main processing from the `ColorView` hosted in the control template to the `ColorPicker` itself. This works-around lifetime limitations with how the control is created (and the template applied".~ Also see:
   * https://github.com/AvaloniaUI/Avalonia/issues/9098#issuecomment-1279804821
   * #9128
 * Updates the sample page in ControlCatalog to include a ColorPicker built from code-behind
 * New ArrayList class (for internal-use only) that avoids an extra copy of List -> Array creating WriteableBitmap
 * Remove the highly specialized `ThirdComponentConverter` replacing it with the new `ColorSpectrum.ThirdComponent` property (which is now automatically calculated based on the spectrum `Components` property). This removes one unnecessarily converter and makes a bit more sense in practical usage.
 * [ColorSlider] Fixes background updates for all property changes
 * [ColorSpectrum] Add missing base() call in constructor
 * [ColorSpectrum] Fix the color spectrum selection ellipse position. Specifically, when color is updated and the control is not part of the visual tree.
 * Additional comments
 * Other minor refactoring and code fixes


## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->


## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->


## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->


## Checklist

- [ ] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- ~[ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation~

## Breaking changes

None

## Obsoletions / Deprecations

None

## Fixed issues

Fixes #9098

robloo committed:
Implement WriteableBitmap caching/reuse and disposal in ColorSlider

robloo committed:
Add new ArrayList to avoid an extra copy when creating spectrum bitmaps

robloo committed:
Fix ColorSlider update/refresh for all property changes

robloo committed:
Make all ColorSpectrum bitmaps globally accessible by the control for future disposal

robloo committed:
Disable WriteableBitmap re-use in ColorSlider due to crashes

robloo committed:
Add missing base constructor call

robloo committed:
Fix the color spectrum selection ellipse position

Specifically, when updated and not part of the visual tree

robloo committed:
Comment additional ColorSpectrum methods

robloo committed:
Fix formatting

robloo committed:
Remove ColorSlider background disposal on attach/detach from visual tree

jp2masa commented on src/Avalonia.Controls.ColorPicker/ColorSpectrum/ColorSpectrum.cs:
```diff
@@ -95,7 +97,7 @@ public partial class ColorSpectrum : TemplatedControl
         /// Initializes a new instance of the <see cref="ColorSpectrum"/> class.
         /// </summary>
-        public ColorSpectrum()
+        public ColorSpectrum() : base()
```
Nit: I don't think this makes any difference, the base constructor is automatically called.

robloo replied:
Yea, true, just done for consistency and habit

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0024464-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Merge branch 'master' into colorpicker-updates-3

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0024664-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Merge branch 'master' into colorpicker-updates-3

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0024770-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Add additional ColorPicker sample to control catalog created from code-behind

robloo committed:
Add and implement IsHostedInColorPicker property to ColorView

This works-around some lifecycle issues by allowing the ColorView to specially handle how it handles property changes based on whether or not it is within a ColorPicker.

robloo commented on src/Avalonia.Controls.ColorPicker/ColorView/ColorView.Properties.cs:
```diff
@@ -395,6 +403,22 @@ public bool IsHexInputVisible
+        /// <see cref="ColorPicker"/>, it does not need to handle most property changes.
+        /// Instead, the primary calculations are done within the <see cref="ColorPicker"/> itself.
+        /// </remarks>
+        public bool IsHostedInColorPicker
```
I'm not a fan of properties like this but it was the lesser of two evils. See the comments in the PR description and here: https://github.com/AvaloniaUI/Avalonia/issues/9098#issuecomment-1279804821

robloo replied:
Nevermind, I slept on it and thought of an obvious simpler solution that doesn't require this 'hacky' property.

robloo replied:
And then it was pointed out adding this property and making the ColorPicker always the primary control for calculations actually fixes setting the color Palette too. So... not sure what to do here and it might change in the future. @maxkatz any comments?

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0024820-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Make some todo-areas of the code more clear

robloo committed:
Remove ThirdComponentConverter replacing it with ColorSpectrum.ThirdComponent property

robloo committed:
Merge branch 'master' into colorpicker-updates-3

robloo commented:
@maxkatz6 This one should be all set for review now. It contains several improvements and two bug fixes.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0024825-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Revert "Add and implement IsHostedInColorPicker property to ColorView"

This reverts commit 2575a10fb09677ac3f131ab3715b20c46b588397.

robloo committed:
Hand-off control to the ColorView after template is applied in ColorPicker

This solution doesn't require the hacky IsHostedInColorPicker property

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0024837-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

maxkatz6 commented on src/Avalonia.Controls.ColorPicker/ColorSpectrum/ColorSpectrum.Properties.cs:
```diff
@@ -218,5 +226,21 @@ public ColorSpectrumShape Shape
+        /// <see cref="ColorComponent"/> type instead of the more accurate <see cref="HsvComponent"/>
+        /// to allow direct usage by the generalized color sliders.
+        /// </remarks>
+        public ColorComponent ThirdComponent
```
If it's supposed to be internal helper-property for the template, I personally like "[TemplateSettings](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.progressbar.templatesettings?view=winrt-22621#windows-ui-xaml-controls-progressbar-templatesettings)" pattern used in UWP.
But I don't think we need to introduce it here just for the one property.

robloo replied:
While this property is for the template right now I wouldn't consider that the only place it could ever be used. It actually fits in nicely with the Components property to make usage of this control with sliders easier -- it makes sense from an API standpoint.

Therefore, I wouldn't use the template settings pattern here at all.

maxkatz6 commented on src/Avalonia.Controls.ColorPicker/ColorSpectrum/ColorSpectrum.Properties.cs:
```diff
@@ -93,6 +93,14 @@ public partial class ColorSpectrum
+        /// <summary>
+        /// Defines the <see cref="ThirdComponent"/> property.
+        /// </summary>
+        public static readonly StyledProperty<ColorComponent> ThirdComponentProperty =
+            AvaloniaProperty.Register<ColorSpectrum, ColorComponent>(
+                nameof(ThirdComponent),
+                ColorComponent.Component3); // Value
```
Just for case, styled property with a private setter can still be modified by the user.
Simply by calling `SetValue(ThirdComponentProperty, value)`, skipping the CLR property call.
Instead, you can make this property a direct property without a setter - it won't be possible to modify the property without hack like reflection.

Something like this:
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/ProgressBar.cs#L26-L30

robloo replied:
Yea, I considered making this a direct property but didn't want to restrict any use cases in the styles or themes. I'm not sure if it would ever be used that way but wanted to allow it. Since it is supposed to be get-only a direct property probably makes sense -- but I don't want to change the API in the future here as a breaking change. So overall decided most powerful is best. 

I also know it can be overridden by calling SetValue() directly. That was actually intentional in case some control downstream ever needs to modify the property. There is still a way to do it. So probably there should be a protected set there too.

maxkatz6 committed:
Merge branch 'master' into colorpicker-updates-3

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0024893-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]
